Services/CustomDeviceModels.cs

```diff
@@ -0,0 +1,222 @@
+            IStorageAdapterClient storage,
+            ILogger logger)
+        {
+            this.storage = storage;
+            this.log = logger;
+        }
+
+        /// <summary>
+        /// Get list of custom device models.
+        /// </summary>
+        public async Task<IEnumerable<DeviceModel>> GetListAsync()
+        {
+            ValueListApiModel data;
+
+            try
+            {
+                data = await this.storage.GetAllAsync(STORAGE_COLLECTION);
+            }
+            catch (Exception e)
+            {
+                this.log.Error("Unable to load device models from storage",
+                    () => new { e.Message, Exception = e });
+                throw new ExternalDependencyException("Unable to load device models from storage", e);
+            }
+
+            try
+            {
+                var results = new List<DeviceModel>();
+                foreach (var item in data.Items)
+                {
+                    var deviceModel = JsonConvert.DeserializeObject<DeviceModel>(item.Data);
+                    deviceModel.ETag = item.ETag;
+                    deviceModel.Type = CUSTOMMODEL;
+                    results.Add(deviceModel);
+                }
+
+                return results;
+            }
+            catch (Exception e)
+            {
+                this.log.Error("Unable to parse device models loaded from storage",
+                    () => new { e.Message, Exception = e });
+                throw new ExternalDependencyException("Unable to parse device models loaded from storage", e);
+            }
+        }
+
+        /// <summary>
+        /// Get a custom device model.
+        /// </summary>
+        public async Task<DeviceModel> GetAsync(string id)
+        {
+            if (string.IsNullOrEmpty(id))
+            {
+                this.log.Error("Device model id cannot be empty!", () => { });
+                throw new InvalidInputException("Device model id cannot be empty! ");
+            }
+
+            var item = await this.storage.GetAsync(STORAGE_COLLECTION, id);
+            var deviceModel = JsonConvert.DeserializeObject<DeviceModel>(item.Data);
+            deviceModel.ETag = item.ETag;
+            return deviceModel;
+        }
+
+        /// <summary>
+        /// Create a custom device model.
+        /// </summary>
+        public async Task<DeviceModel> InsertAsync(DeviceModel deviceModel, bool generateId = true)
+        {
+            deviceModel.Created = DateTimeOffset.UtcNow;
+            deviceModel.Modified = deviceModel.Created;
+
+            if (generateId)
+            {
+                deviceModel.Id = Guid.NewGuid().ToString();
+            }
+
+            this.log.Debug("Creating a custom device model.", () => new { deviceModel });
+
+            try
+            {
+                // Note: using UpdateAsync because the service generates the ID
+                var result = await this.storage.UpdateAsync(
```

Comment:
the call to UpdateAsync is not tested completely:
* verify that the JSON written to storage contains the right data
* verify that the device model ID is used

Reply:
Done.

Reply:
I can't see this test, could you point me to it? is the PR up to date?